### PR TITLE
Added FCM push notifications

### PR DESCRIPTION
**Pull request by TheHollidayInn**
Fixes put_issue_url_here
### Changes

This adds FCM push notifications to the push notification library. This was suggested by Google to upgrade to Firebase.

---

UUID:

### CODE REVIEW

File: website/server/libs/api-v3/pushNotifications.js [hunk 1]
```diff
@@ -92,6 +96,25 @@ module.exports = function sendNotification (user, details = {}) {
             data: payload,
           });
         }
+
+        if (fcm) {
+          var message = {
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: let

File: website/server/libs/api-v3/pushNotifications.js [hunk 2]
```diff
@@ -92,6 +96,25 @@ module.exports = function sendNotification (user, details = {}) {
...
+              data: payload,
+              notification: {
+                  title: details.title,
+                  body: details.message,
+              }
+          };
+
+          fcm.send(message, function(err, response){
+              if (err) {
+                  console.log(err, "Something has gone wrong!");
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: should use the logger module

File: website/server/libs/api-v3/pushNotifications.js [hunk 3]
```diff
@@ -92,6 +96,25 @@ module.exports = function sendNotification (user, details = {}) {
...
+              } else {
+                  console.log("Successfully sent with response: ", response);
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: this should be removed, if the notification is sent correctly this isn't really necessary

**Comment by paglias** [comment]
I see some problems here:
- registration ids used for gcm can be used also for fcm?
- the code right now will send the notification twice, once for GCM and once for FCM

also there are some conflicts

**Comment by paglias** [comment]
Also it looks like our current library uses node-gcm that supports FCM too https://github.com/ToothlessGear/node-gcm/issues/255. But push-notify doesn't currently have the latest version of node-gcm

File: website/server/libs/api-v3/pushNotifications.js (hunk 2 above)
Comment (TheHollidayInn) [inline website/server/libs/api-v3/pushNotifications.js]: Cool, I wasn't sure where to log this information.

**Comment by TheHollidayInn** [comment]
Yes, regIds work for both. And, I think I should remove the GCM code since there is nothing implemented to use it on the Android side.

File: test/api/v3/unit/libs/pushNotifications.js [hunk 4]
```diff
@@ -17,9 +19,12 @@ describe('pushNotifications', () => {
   beforeEach(() => {
     user = new User();
     gcmSendSpy = sinon.spy();
+    fcmSendSpy = sinon.spy();
     apnSendSpy = sinon.spy();
 
-    sandbox.stub(nconf, 'get').returns('true');
+    sandbox.stub(nconf, 'get').returns('true-key');
+
+    sandbox.stub(FCM.prototype, "send", fcmSendSpy)
```
Comment (khipkin) [inline test/api/v3/unit/libs/pushNotifications.js]: missing semicolon

**Comment by TheHollidayInn** [comment]
Updated!

File: package.json [hunk 5]
```diff
@@ -31,6 +31,9 @@
     "express": "~4.13.3",
     "express-csv": "~0.6.0",
     "express-validator": "^2.18.0",
+    "fcm-push": "^1.0.7",
+    "firebase": "^2.2.9",
```
Comment (paglias) [inline package.json]: Are these firebase modules necessary?

File: website/server/libs/api-v3/pushNotifications.js [hunk 6]
```diff
@@ -7,8 +7,12 @@ import Bluebird from 'bluebird';
...
   S3,
 } from './aws';
+import FCM from 'fcm-push';
 
 const GCM_API_KEY = nconf.get('PUSH_CONFIGS:GCM_SERVER_API_KEY');
+const FCM_API_KEY = nconf.get('PUSH_CONFIGS:FCM_SERVER_API_KEY');
+
+let fcm = FCM_API_KEY ? new FCM(FCM_API_KEY) : undefined;
 
 let gcm = GCM_API_KEY ? pushNotify.gcm({
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: gcm is still being setup here even if it's not used

**Comment by TheHollidayInn** [comment]
Updated.

File: test/api/v3/unit/libs/pushNotifications.js [hunk 7]
```diff
@@ -40,14 +45,16 @@ describe('pushNotifications', () => {
 
   it('throws if user is not supplied', () => {
     expect(sendPushNotification).to.throw;
-    expect(gcmSendSpy).to.not.have.been.called;
+    // expect(gcmSendSpy).to.not.have.been.called;
```
Comment (crookedneighbor) [inline test/api/v3/unit/libs/pushNotifications.js]: Why keep this around?

File: website/server/libs/api-v3/pushNotifications.js [hunk 8]
```diff
@@ -7,19 +7,23 @@ import Bluebird from 'bluebird';
...
-  retries: 3,
-}) : undefined;
+let fcm = FCM_API_KEY ? new FCM(FCM_API_KEY) : undefined;
 
-if (gcm) {
-  gcm.on('transmissionError', (err, message, registrationId) => {
-    logger.error('GCM Error', err, message, registrationId);
-  });
-}
+// let gcm = GCM_API_KEY ? pushNotify.gcm({
```
Comment (crookedneighbor) [inline website/server/libs/api-v3/pushNotifications.js]: Why are we keeping this code here?

Comment (TheHollidayInn) [inline website/server/libs/api-v3/pushNotifications.js]: @paglias mentioned above that the push notify library supports FCM but doesn't have the update version of node-gcm. So, I figured if we would leave this code incase they update. I also suggested removing it - so either way is fine with me.

**Comment by TheHollidayInn** [comment]
Updated

**Comment by TheHollidayInn** [comment]
Updated

**Pull request closed without merging**